wgpu-core/src/device/queue.rs

```diff
@@ -458,17 +467,133 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         Ok(())
     }
 
+    // Enacts all zero initializations required by the given command buffers
+    // Required commands are appended to device.pending_writes
+    fn initialize_used_uninitialized_memory<B: GfxBackend>(
+        &self,
+        queue_id: id::QueueId,
+        command_buffer_ids: &[id::CommandBufferId],
+    ) -> Result<(), QueueSubmitError> {
+        if command_buffer_ids.is_empty() {
+            return Ok(());
+        }
+
+        let hub = B::hub(self);
+        let mut token = Token::root();
+
+        let mut required_buffer_inits = {
+            let (command_buffer_guard, mut token) = hub.command_buffers.read(&mut token);
+            let (mut buffer_guard, _) = hub.buffers.write(&mut token);
```

Comment:
Could we perhaps move this check down into the inner loop, and only after checking that `cmdbuf.buffer_memory_init_actions` is non-empty?